Add work-around to get printing in OOP frames to work in Chrome

Rob--W opened the pull request:
Edit: Moved part of the refactoring and fix for #7720 to #7721.

---

Fixes #7706 

I suggest to review the three commits separately, since they are kind of independent.

@yurydelendik Can you take a look at the first two commits? They are a kind of follow-up to your previous patch at #7697.

About half of the third commit are comments to explain what it does. Its commit message also includes steps for testing. @timvandermeij Could you review the third commit?

Rob--W committed:
Strictly manage lifetime of PDFPrintService

Make sure that the print service is stopped as soon as possible when
aborted, and that it is not possible for a (slow) promise to
accidentally wipe the state of a print job that was started later.

Rob--W committed:
Close overlay if print service was not initialized

Fixes #7720

Rob--W committed:
Add work-around for printing from OOP frames

To test:

1. Start Chrome with `--isolate-extensions` (if using Chrome 54-).
2. Load the extension.
3. Follow the steps from https://crbug.com/650952.

To see all relevant debug/console messages:

- Right-click inside the PDF viewer, inspect element.
- Open devtools for the tab.

Known limitations:

- Print does not start if extension cannot run scripts in the top frame.
- May trigger mixed content console warnings: https://crbug.com/437440
- If top page acts hostile (e.g. removes frame), then the print does not
  start (but no information is leaked!).

Snuffleupagus commented on web/pdf_print_service.js:
```diff
@@ -200,7 +211,17 @@
+      activeService.renderPages().then(function() {
+        activeServiceOnEntry.throwIfInactive();
+        return startPrint(activeServiceOnEntry);
+      }).then(null, function() {
```
Nit: I may very well be overlooking something here, but couldn't you just use `catch` here instead of these two `then`s? I.e. the code below simply becomes:
```javascript
}).catch(function() {
  if (activeServiceOnEntry.isActive()) {
    abort();
  }
});
```

Rob--W replied:
Fixed.

Snuffleupagus commented on web/chrome_print_service_extension.js:
```diff
@@ -0,0 +1,477 @@
+    window.addEventListener('message', onmessage, true);
+    function onmessage(event) {
+      if (event.origin === EXTENSION_ORIGIN && event.data === nonceNoDupe) {
+        LogStep(4, 'Receive message from X');
```
Travis complains about these, you probably need to add `LogStep` to the list of `globals`.

Snuffleupagus commented on web/chrome_print_service_extension.js:
```diff
@@ -0,0 +1,477 @@
+
+  // This is only for debugging. The function is also serialized and passed to
+  // the content script (CS), which also passes it to frame F.
+  function LogStep(step, description) {
```
Nit: Travis is complaining everywhere that this function is called, `Missing 'new' prefix when invoking a constructor.`, so this may need to be named `logStep` instead.

Rob--W replied:
Done.

Rob--W commented on web/pdf_print_service.js:
```diff
@@ -181,6 +182,16 @@
       return new Promise(renderNextPage);
     },
+
+    isActive: function() {
```
Nope, I just changed it to a getter.

timvandermeij commented:
There are some linting failures. Could you resolve those?

Rob--W commented:
@timvandermeij Done. Please take a look.

The `topContentScript` and `forwardPrintResults` implement the non-trivial Chrome extension communication logic, whereas `processPrintJob` and anything below `getStyleSheetTextForPrinting` implements print/PDF.js. I'm sure that you're able to review the second part, but if you have difficulty with going through the first part I would totally understand (hopefully the ~200 lines of explanatory comments clarifies things a little bit).

yurydelendik commented on web/pdf_print_service.js:
```diff
@@ -200,7 +211,17 @@
+        // Ignore any error messages.
+      }).then(function() {
+        if (activeServiceOnEntry.active) {
+          abort();
```
I'm a little bit confused about why we need to abort when the service is active.

Rob--W replied:
If somehow the active service has been swapped for another service, then we don't want to abort since the service works on the active service.

yurydelendik replied:
That's what I thought, but `if (activeServiceOnEntry.active) abort()` kinda says otherwise?

Rob--W replied:
`activeServiceOnEntry.active` is true if the currently active service is still the same as the service from when the print task started.

If somehow the task was aborted early (e.g. abort was called early), then the "global" activeService is null. Now the interesting case is when another print job is started immediately after abortion of the previous task. Then the `if (activeService)` condition is true, but the async call to abort from the original print request actually cancels the second request instead of the first request (which was already cancelled in the first place).

yurydelendik replied:
So `activeServiceOnEntry == activeSerice` and we are calling abort() which cancels activeService a.k.a. activeServiceOnEntry -- that's still confuses me. Can we change the code so it will be read better, e.g. move into the startTask method or add more readable check (not sure if more comments are suitable here)?

yurydelendik replied:
Ignore the above. Is `startPrint` actually `performPrint` -- I read startPrint as something that initiating print?

Rob--W replied:
`startPrint` starts the actual printing, so yes, it is also `performPrint`, or `startPrintForReal`. I will add a comment that explains the check.

yurydelendik reviewed (commented):
First two commits looks good (with comment addressed)

I don't think having `PDFPrintService.delegate`/handler will be a good practice. Can you refactor code further to move image creation into standard "PDFPrintServiceRenderer" (?) and it can be override by chrome print service when needed? (In another words, I want to avoid `if (handled) {...`.)

I was not reviewing "web/chrome_print_service_extension.js".

Rob--W commented:
@yurydelendik The default renderer should still be used in the Chrome extension (e.g. in top-level frames or in-process frames).

Are you thinking of something like this (for brevity I'm using ES6 syntax):

```javascript
// in pdf_print_service.js
exports.PDFPrintServiceRenderer = {
  initService() {},
  destroyService() {},
  handlePage(...) {
    // existing implementation using img/toBlob.
  },
  startPrint(...) {
    // existing implementation of startPrint
  }
};
```

```javascript
// in chrome_print_service_extension.js
// pdfPrintService = exports from pdf_print_service.js
var {handlePage, startPrint} = pdfPrintService.PDFPrintServiceRenderer;
pdfPrintService.PDFPrintServiceRenderer = {
  handlePage(printService, img, canvas) {
    if (isOutOfProcessFrame()) {
      return /* OOP-specific logic here */;
    }
    // Default action:
    return handlePage(printService, img, canvas);
  },
  // Similarly for startPrint. Also initService and destroyService here.
};
```

yurydelendik commented:
Correct. The above "PDFPrintServiceRenderer" aligns with what I'm thinking. (Maybe make it more non-inlined: default is `PDFPrintService.renderer = new PDFPrintServiceRenderer();` and for Chrome `PDFPrintService.renderer = new OOPPrintServiceRenderer();`?)

Rob--W commented:
@yurydelendik 

Where would you put that assignment? It can't be in the implementation files, because the default renderer is also used in the Chrome -specific adapter.

Rob--W commented:
I can also return early and not overwrite anything at all if I detect that the extension is in an oop frame.

But still both modules need to be loaded, how do I guarantee that the Chrome-specific module loads after the default one?

yurydelendik commented:
> But still both modules need to be loaded, how do I guarantee that the Chrome-specific module loads after the default one?

You import the export from first/default module -- it will guarantee the loading of the latter.

yurydelendik commented:
> Where would you put that assignment? It can't be in the implementation files, because the default renderer is also used in the Chrome -specific adapter.

pdf_print_service.js: `PDFPrintService.renderer = new PDFPrintServiceRenderer();`

chrome_print_service.js: import PDFPrintService from 'pdf_print_service'; ... `PDFPrintService.renderer = new OOPPrintServiceRenderer();`

yurydelendik commented:
Just to notice: you are already have almost exact`PDFPrintService.renderer` which is `PDFPrintService.delegate`, but I want it to take more logic. We may go one step further and make it `PDFPrintService.rendererFactory` and instantiate it only when service is active.

Rob--W committed:
Refactor page printing logic on the web

- Move the global scratchCanvas to PDFPrintService. This is mainly to
  make it easier to reason about the state of scratchCanvas. In practice
  there is no difference because only one PDFPrintService instance can
  be instantiated at any given time.

- Move all logic of using the rendered page to one location.
  This makes it easier to replace the printing logic later, when I add
  special handling to out-of-process frames in the Chrome extension.

Rob--W committed:
Refactor printing: startPrint -> performPrint

- Renamed startPrint to performPrint to emphasize that the method
  does not start the print process (preparing pages for the printer),
  but that it does the actual printing (sending pages off to the
  printer).

- Put performPrint in the PDFPrintService, so that it can be
  overridden if needed.

Rob--W commented:
@yurydelendik I slightly refactored the printing logic, can you take a look at the third and fourth commit (the ones starting with "Refactor")? See their commit messages.

In the end I did not introduce any factory or delegate, I simply override methods of PDFPrintService.prototype from chrome_print_service_extension.js

yurydelendik reviewed (approved):
General printing logic looks good. I did not review the code in chrome print server, however comments are reasonable (assuming notes are correct).

yurydelendik commented:
/botio-windows preview

timvandermeij commented:
Closing as fixed upstream.

Sofiane000 commented:
@timvandermeij 
I implemented the example Hello world using base64 : https://jsfiddle.net/pdfjs/cq0asLqz/
I am trying to add a print button to it.
I am looking into pdf_print_service.js but I did not know how to incorporate it in my example.
Do I need to use the pdf viewer and if yes how would I do that?
Thanks
